File: java/src/main/java/com/cc/cli/TicTacToeCli.java
package com.cc.cli;

import com.cc.tictactoe.GameStatus;
import com.cc.tictactoe.TicTacToe;
import com.cc.tictactoe.TicTacToeElement;

import java.util.Scanner;


public class TicTacToeCli extends TicTacToe {
    private final Scanner scanner = new Scanner(System.in);

    public TicTacToeCli(TicTacToeElement turn) {
        super(turn);
    }

    public void printBoard() {
        TicTacToeElement[][] board = getBoard();
        String rowSeparator = "--+---+--";
        for (int i = 0; i < board.length; i++) {
            StringBuilder row = new StringBuilder();
            for (int j = 0; j < board[i].length; j++) {
                row.append(board[i][j] == null ? " " : board[i][j].toString());
                if (j < board[i].length - 1) {
                    row.append(" | ");
                }
            }
            System.out.println(row);
            if (i < board.length - 1) {
                System.out.println(rowSeparator);
            }
        }
        System.out.println();
    }

    public void printTurn() {
        TicTacToeElement turn = getTurn();
        System.out.println("Turn: " + turn);
    }

    public void inputFromCli() {
        try {
            System.out.println("which cell you want to input?");
            System.out.print("row (0-2): ");
            int rowIndex = scanner.nextInt();
            System.out.print("column (0-2): ");
            int columnIndex = scanner.nextInt();
            boolean success = input(rowIndex, columnIndex);
            if (!success) {
                System.out.println("Invalid input");
            }
        } catch (Exception e) {
            System.out.println("Invalid input");
            scanner.nextLine();
        }
    }

    public boolean isInProgress() {
        GameStatus status = getGameStatus();
        switch (status) {
            case DRAW -> {
                System.out.println("Draw!");
                return false;
            }
            case X_WINS -> {
                System.out.println("X wins!");
                printBoard();
                return false;
            }
            case O_WINS -> {
                System.out.println("O wins!");
                printBoard();
                return false;
            }
            case IN_PROGRESS -> {
                return true;
            }
            default -> throw new IllegalStateException("Unexpected value: " + status);
        }
    }

    public boolean resetBoardFromCli() {
        try {
            System.out.println("Play again? (y/n)");
            String response = scanner.next();
            switch (response) {
                case "y" -> {
                    TicTacToeElement newTurn = getStartTurn() == TicTacToeElement.X ? TicTacToeElement.O :
                            TicTacToeElement.X;
                    resetBoard(newTurn);
                    System.out.println("New game started with " + newTurn + "first");
                    return true;
                }
                case "n" -> {
                    return false;
                }
                default -> {
                    System.out.println("Invalid input");
                    return resetBoardFromCli();
                }
            }
        } catch (Exception e) {
            System.out.println("Invalid input");
            scanner.nextLine();
            return resetBoardFromCli();
        }
    }
}

File: ts/src/game-in-cli/TicTacToeCli.ts
import TicTacToe from "../TicTacToe/TicTacToe";
import { TicTacToeElement, GameStatus } from "../TicTacToe/typing";
import { scanner } from "./Scanner";

export class TicTacToeCli extends TicTacToe {
  constructor(turn: TicTacToeElement) {
    super(turn);
  }

  public printBoard(): void {
    const board = this.getBoard();
    const rowSeparator = "--+---+--";
    for (let i = 0; i < board.length; i++) {
      let row = "";
      for (let j = 0; j < board[i].length; j++) {
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        row += board[i][j] === null ? " " : board[i][j]!.toString();
        if (j < board[i].length - 1) {
          row += " | ";
        }
      }
      console.log(row);
      if (i < board.length - 1) {
        console.log(rowSeparator);
      }
    }
  }

  public printTurn(): void {
    const turn: TicTacToeElement = this.getTurn();
    console.log("Turn: " + turn);
  }

  public async inputFromCli(): Promise<void> {
    try {
      const rowIndex: number = await scanner.nextInt("row (0-2): ");
      const columnIndex: number = await scanner.nextInt("column (0-2): ");
      const success: boolean = this.input(rowIndex, columnIndex);
      if (!success) {
        console.log("Invalid input");
        await this.inputFromCli();
      }
    } catch (e) {
      console.log("Invalid input");
      await this.inputFromCli();
    }
  }

  public isInProgress(): boolean {
    const status: GameStatus = this.getGameStatus();
    switch (status) {
      case GameStatus.DRAW:
        console.log("Draw!");
        return false;
      case GameStatus.X_WINS:
        console.log("X wins!");
        this.printBoard();
        return false;
      case GameStatus.O_WINS:
        console.log("O wins!");
        this.printBoard();
        return false;
      case GameStatus.IN_PROGRESS:
        return true;
    }
  }

  public async resetBoardFromCli(): Promise<boolean> {
    try {
      const response: string = await scanner.next("Play again? (y/n) ");
      switch (response) {
        case "y": {
          const newTurn: TicTacToeElement =
            this.getStartTurn() == TicTacToeElement.X ? TicTacToeElement.O : TicTacToeElement.X;
          this.resetBoard(newTurn);
          console.log(`New game started with ${newTurn} first`);
          return true;
        }
        case "n":
          return false;
        default:
          console.log("Invalid input");
          return await this.resetBoardFromCli();
      }
    } catch (e) {
      console.log("Invalid input");
      return await this.resetBoardFromCli();
    }
  }
}
